cmd/tsconnect/src/ssh.tsx

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+import { useState, useCallback } from "preact/hooks"
+import { Terminal } from "xterm"
+import { FitAddon } from "xterm-addon-fit"
+
+type SSHSessionDef = {
+  username: string
+  hostname: string
+}
+
+export function SSH({ netMap, ipn }: { netMap: IPNNetMap; ipn: IPN }) {
+  const [sshSessionDef, setSSHSessionDef] = useState<SSHSessionDef | null>(null)
+  const clearSSHSessionDef = useCallback(() => setSSHSessionDef(null), [])
+  if (sshSessionDef) {
+    return (
+      <SSHSession def={sshSessionDef} ipn={ipn} onDone={clearSSHSessionDef} />
+    )
+  }
+  const sshPeers = netMap.peers.filter(
+    (p) => p.tailscaleSSHEnabled && p.online !== false
+  )
+
+  if (sshPeers.length == 0) {
+    return <NoSSHPeers />
+  }
+
+  return <SSHForm sshPeers={sshPeers} onSubmit={setSSHSessionDef} />
+}
+
+function SSHSession({
+  def,
+  ipn,
+  onDone,
+}: {
+  def: SSHSessionDef
+  ipn: IPN
+  onDone: () => void
+}) {
+  return (
+    <div
+      class="flex-grow bg-black p-2 overflow-hidden"
+      ref={(node) => {
```

Comment:
It doesn't look like the function is returning anything, what is the ref handler holding onto?

Reply:
I'm using `ref` as a callback to know when the component has been rendered and there's a DOM node I can attach the xterm.js instance to. If there's a more idiomatic way with to do with hooks I'm happy to switch (we never did a full hooks port at my old job, so I don't have much production experience with them).

Reply:
Ohhh that's clever. The more idiomatic way would be to combine a `useRef` on this div with a `useEffect` that has that ref in the dependency array, like what we're doing with the [ACL editor](https://github.com/tailscale/corp/blob/main/control/adminhttp/panel/src/components/acl-editor/acl-editor.tsx#L56) but I think the logic is the same in the end. Up to you if you want to switch!

Reply:
Keeping this for now mostly so I can merge the commit as is (don't have a dev environment on my vacation machine, but I don't want to have this PR block Charlotte's work by having that be done on top of the old DOM code). Will look into switching to hooks next week.